File: web/src/lib/api/inventory.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { apiFetch } from "@/lib/api";

// ── Types ─────────────────────────────────────────────────────────────────────

export interface Item {
  id: string;
  sku: string;
  name: string;
  description?: string;
  unit: string;
  category?: string;
  min_stock_qty?: string;
  attributes: Record<string, unknown>;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface Location {
  id: string;
  code: string;
  name: string;
  parent_id?: string;
  type: string;
  is_active: boolean;
}

export interface Lot {
  id: string;
  item_id: string;
  lot_number: string;
  supplier_ref?: string;
  received_at: string;
  expires_at?: string;
  status: string;
}

export interface StockLevel {
  lot_id: string;
  location_id: string;
  quantity: string;
  item_id: string;
  item_name: string;
  item_sku: string;
  unit: string;
  lot_number: string;
  location_code: string;
  location_name: string;
}

export interface StockMovement {
  id: number;
  lot_id: string;
  location_id: string;
  quantity: string;
  movement_type: string;
  unit_cost?: string;
  notes?: string;
  actor_id: string;
  created_at: string;
}

// ── API functions ─────────────────────────────────────────────────────────────

export const inventoryApi = {
  listItems: (params?: { q?: string; category?: string; limit?: number; offset?: number }) =>
    apiFetch<{ items: Item[] }>(`/api/items?${new URLSearchParams(
      Object.entries(params ?? {}).filter(([, v]) => v != null).map(([k, v]) => [k, String(v)])
    )}`),

  getItem: (id: string) => apiFetch<Item>(`/api/items/${id}`),

  createItem: (body: {
    sku: string; name: string; unit: string;
    description?: string; category?: string; min_stock_qty?: string;
  }) => apiFetch<Item>("/api/items", { method: "POST", body: JSON.stringify(body) }),

  updateItem: (id: string, body: Partial<Pick<Item, "name" | "description" | "unit" | "category" | "min_stock_qty">>) =>
    apiFetch<Item>(`/api/items/${id}`, { method: "PATCH", body: JSON.stringify(body) }),

  deactivateItem: (id: string) =>
    apiFetch<void>(`/api/items/${id}`, { method: "DELETE" }),

  listLocations: () => apiFetch<{ locations: Location[] }>("/api/locations"),

  createLocation: (body: { code: string; name: string; type?: string; parent_id?: string }) =>
    apiFetch<Location>("/api/locations", { method: "POST", body: JSON.stringify(body) }),

  listLots: (params?: { item_id?: string; limit?: number; offset?: number }) =>
    apiFetch<{ lots: Lot[] }>(`/api/lots?${new URLSearchParams(
      Object.entries(params ?? {}).filter(([, v]) => v != null).map(([k, v]) => [k, String(v)])
    )}`),

  createLot: (body: { item_id: string; lot_number: string; supplier_ref?: string; expires_at?: string }) =>
    apiFetch<Lot>("/api/lots", { method: "POST", body: JSON.stringify(body) }),

  getCurrentStock: (params?: { item_id?: string }) =>
    apiFetch<{ stock: StockLevel[] }>(`/api/stock/current${params?.item_id ? `?item_id=${params.item_id}` : ""}`),

  listMovements: (params?: { lot_id?: string; limit?: number }) =>
    apiFetch<{ movements: StockMovement[] }>(`/api/stock/movements?${new URLSearchParams(
      Object.entries(params ?? {}).filter(([, v]) => v != null).map(([k, v]) => [k, String(v)])
    )}`),

  recordMovement: (body: {
    lot_id: string; location_id: string; quantity: string;
    movement_type: string; idempotency_key: string; unit_cost?: string; notes?: string;
  }) => apiFetch<StockMovement>("/api/stock/movements", { method: "POST", body: JSON.stringify(body) }),
};

// ── Hooks ─────────────────────────────────────────────────────────────────────

export const useItems = (params?: Parameters<typeof inventoryApi.listItems>[0]) =>
  useQuery({
    queryKey: ["items", params],
    queryFn: () => inventoryApi.listItems(params),
    select: (d) => d.items,
  });

export const useItem = (id: string) =>
  useQuery({
    queryKey: ["items", id],
    queryFn: () => inventoryApi.getItem(id),
    enabled: !!id,
  });

export const useCreateItem = () => {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: inventoryApi.createItem,
    onSuccess: () => qc.invalidateQueries({ queryKey: ["items"] }),
  });
};

export const useLocations = () =>
  useQuery({
    queryKey: ["locations"],
    queryFn: inventoryApi.listLocations,
    select: (d) => d.locations,
  });

export const useCreateLocation = () => {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: inventoryApi.createLocation,
    onSuccess: () => qc.invalidateQueries({ queryKey: ["locations"] }),
  });
};

export const useLots = (params?: Parameters<typeof inventoryApi.listLots>[0]) =>
  useQuery({
    queryKey: ["lots", params],
    queryFn: () => inventoryApi.listLots(params),
    select: (d) => d.lots,
  });

export const useCurrentStock = (params?: { item_id?: string }) =>
  useQuery({
    queryKey: ["stock", "current", params],
    queryFn: () => inventoryApi.getCurrentStock(params),
    select: (d) => d.stock,
  });

export const useRecordMovement = () => {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: inventoryApi.recordMovement,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["stock"] });
      qc.invalidateQueries({ queryKey: ["lots"] });
    },
  });
};

File: internal/inventory/inventory.go
package inventory

import (
	"encoding/json"
	"errors"
	"time"

	"github.com/shopspring/decimal"
)

var (
	ErrItemNotFound      = errors.New("item not found")
	ErrLocationNotFound  = errors.New("location not found")
	ErrLotNotFound       = errors.New("lot not found")
	ErrDuplicateSKU      = errors.New("SKU already exists")
	ErrDuplicateLot      = errors.New("lot number already exists for this item")
	ErrDuplicateMovement = errors.New("movement with this idempotency key already exists")
)

const (
	MovementReceipt    = "receipt"
	MovementAdjustment = "adjustment"
	MovementTransfer   = "transfer"
)

type Item struct {
	ID          string           `json:"id"`
	SKU         string           `json:"sku"`
	Name        string           `json:"name"`
	Description *string          `json:"description,omitempty"`
	Unit        string           `json:"unit"`
	Category    *string          `json:"category,omitempty"`
	MinStockQty *decimal.Decimal `json:"min_stock_qty,omitempty"`
	Attributes  json.RawMessage  `json:"attributes"`
	IsActive    bool             `json:"is_active"`
	CreatedAt   time.Time        `json:"created_at"`
	UpdatedAt   time.Time        `json:"updated_at"`
}

type Location struct {
	ID       string  `json:"id"`
	Code     string  `json:"code"`
	Name     string  `json:"name"`
	ParentID *string `json:"parent_id,omitempty"`
	Type     string  `json:"type"`
	IsActive bool    `json:"is_active"`
}

type Lot struct {
	ID          string          `json:"id"`
	ItemID      string          `json:"item_id"`
	LotNumber   string          `json:"lot_number"`
	SupplierRef *string         `json:"supplier_ref,omitempty"`
	ReceivedAt  time.Time       `json:"received_at"`
	ExpiresAt   *time.Time      `json:"expires_at,omitempty"`
	Status      string          `json:"status"`
	Attributes  json.RawMessage `json:"attributes"`
}

type StockMovement struct {
	ID           int64            `json:"id"`
	LotID        string           `json:"lot_id"`
	LocationID   string           `json:"location_id"`
	Quantity     decimal.Decimal  `json:"quantity"`
	MovementType string           `json:"movement_type"`
	ReferenceID  *string          `json:"reference_id,omitempty"`
	UnitCost     *decimal.Decimal `json:"unit_cost,omitempty"`
	Notes        *string          `json:"notes,omitempty"`
	ActorID      string           `json:"actor_id"`
	CreatedAt    time.Time        `json:"created_at"`
}

type StockLevel struct {
	LotID        string          `json:"lot_id"`
	LocationID   string          `json:"location_id"`
	Quantity     decimal.Decimal `json:"quantity"`
	ItemID       string          `json:"item_id"`
	ItemName     string          `json:"item_name"`
	ItemSKU      string          `json:"item_sku"`
	Unit         string          `json:"unit"`
	LotNumber    string          `json:"lot_number"`
	LocationCode string          `json:"location_code"`
	LocationName string          `json:"location_name"`
}

// Request types

type CreateItemRequest struct {
	SKU         string           `json:"sku"`
	Name        string           `json:"name"`
	Description *string          `json:"description"`
	Unit        string           `json:"unit"`
	Category    *string          `json:"category"`
	MinStockQty *decimal.Decimal `json:"min_stock_qty"`
	Attributes  json.RawMessage  `json:"attributes"`
}

type UpdateItemRequest struct {
	Name        *string          `json:"name"`
	Description *string          `json:"description"`
	Unit        *string          `json:"unit"`
	Category    *string          `json:"category"`
	MinStockQty *decimal.Decimal `json:"min_stock_qty"`
	Attributes  json.RawMessage  `json:"attributes"`
}

type CreateLocationRequest struct {
	Code     string  `json:"code"`
	Name     string  `json:"name"`
	ParentID *string `json:"parent_id"`
	Type     string  `json:"type"`
}

type CreateLotRequest struct {
	ItemID      string          `json:"item_id"`
	LotNumber   string          `json:"lot_number"`
	SupplierRef *string         `json:"supplier_ref"`
	ReceivedAt  *time.Time      `json:"received_at"`
	ExpiresAt   *time.Time      `json:"expires_at"`
	Attributes  json.RawMessage `json:"attributes"`
}

type RecordMovementRequest struct {
	LotID          string           `json:"lot_id"`
	LocationID     string           `json:"location_id"`
	Quantity       decimal.Decimal  `json:"quantity"`
	MovementType   string           `json:"movement_type"`
	ReferenceID    *string          `json:"reference_id"`
	IdempotencyKey string           `json:"idempotency_key"`
	UnitCost       *decimal.Decimal `json:"unit_cost"`
	Notes          *string          `json:"notes"`
}

type TransferRequest struct {
	LotID            string          `json:"lot_id"`
	SourceLocationID string          `json:"source_location_id"`
	DestLocationID   string          `json:"dest_location_id"`
	Quantity         decimal.Decimal `json:"quantity"`
	IdempotencyKey   string          `json:"idempotency_key"`
	Notes            *string         `json:"notes"`
}
